### PR TITLE
Make sure EventCollection doesn't modify input in-place

### LINKED ISSUE #14487: eventplot sorts np.array positions, but not list positions
### Bug report

**Bug summary**

When using eventplot, the data is sorted in-place if it's passed as a numpy array. If the data is passed as a list, this doesn't happen. 

**Code for reproduction**

```python
from matplotlib import pyplot as plt
import numpy as np

# This sorts the data
sample_data = np.array([3, 2, 1, 10])
plt.eventplot(sample_data)
print(sample_data)

# This does not sort the data
sample_data = [3, 2, 1, 10]
plt.eventplot(sample_data)
print(sample_data)
```

**Actual outcome**

```
[ 1  2  3 10]
[3, 2, 1, 10]

```

**Expected outcome**
For consistency, it would be good if the outcome was always sorted, or always untouched. My preference would be that the data remains unsorted.

**Matplotlib version**

  * Operating system: Ubuntu 16.04.4
  * Matplotlib version: 3.0.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.7.2
  * Jupyter version (if applicable): jupyterlab version 0.35.4
  * Other libraries: numpy version 1.16.2

All libraries were installed using conda.

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #14487. Not entirely sure if casting any input to an array is correct, but will see if the tests all pass.

### CODE REVIEW

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -1462,7 +1462,7 @@ def __init__(self,
 
         .. plot:: gallery/lines_bars_and_markers/eventcollection_demo.py
         """
-
+        positions = np.array(positions, copy=True)
         segment = (lineoffset + linelength / 2.,
                    lineoffset - linelength / 2.)
         if positions is None or len(positions) == 0:
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: At least this line suggests it was intended that `positions` could be None, but this PR changes the handling of None.

Comment (dstansby) [reply lib/matplotlib/collections.py]: Huh, you're right, even if there is a comment above (https://github.com/matplotlib/matplotlib/pull/14488/files#diff-506c6bd01694bddbd8999f2c6e920705R1408) that implies `positions` can't be `None`. I'll fix it to keep the current behaviour with None though (and add a smoke test for `positions=None`)

Comment (anntzer) [reply lib/matplotlib/collections.py]: I don't really know if it can be None or not, could probably use some cleanup :p

Comment (dstansby) [reply lib/matplotlib/collections.py]: Since passing `positions=None` would fail later on in some cases on `positions.sort()`, I've explicitly disallowed `None` input.

Comment (anntzer) [reply lib/matplotlib/collections.py]: It would previously not fail because the sort() is in another else clause; indeed `gca().add_collection(EventCollection(None))` works.
On the other hand there's basically not benefit to supporting None as input (in particular `eventplot(None)` already doesn't work) so I'm fine with dropping that, won't even insist on an api changes note... :)

File: lib/matplotlib/collections.py [hunk 2]
```diff
@@ -1462,12 +1462,14 @@ def __init__(self,
 
         .. plot:: gallery/lines_bars_and_markers/eventcollection_demo.py
         """
-
+        if positions is None:
+            raise ValueError('positions must be an array-like object')
+        positions = np.array(positions, copy=True)
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: Technically np.array already copies by default (and there's likely a ton of places in the codebase which rely on that), so just a comment `# force a copy` is enough.

Comment (dstansby) [reply lib/matplotlib/collections.py]: I think having `copy=True` makes it much more clear what is going on, but happy to add a comment too

**Comment by anntzer** [comment]
anyone can merge post ci

**Comment by QuLogic** [comment]
@meeseeksdev backport to v3.1.x
